nifpga/bitfile.py

```diff
@@ -75,27 +84,480 @@ def base_address_on_device(self):
+        packed_data = packed_data << self._size_in_bits
+        return packed_data | (data_to_pack & self._data_mask)
+
+
+class _Float(object):
+    """ Handles packing and unpacking floating point values from the FPGA. """
+    def __init__(self, name, type_name):
+        self._name = name
+        if "SGL" == type_name:
+            self._size_in_bits = 32
+            self._datatype = DataType.Sgl
+        elif "DBL" == type_name:
+            self._size_in_bits = 64
+            self._datatype = DataType.Dbl
+        self._data_mask = (1 << self._size_in_bits) - 1
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._datatype
+
+    @property
+    def size_in_bits(self):
+        return self._size_in_bits
+
+    @property
+    def is_c_api_type(self):
         return True
 
-    def _is_fifo_fxp(self, channel_xml):
-        datatype = channel_xml.find("DataType").find("SubType").text.title()
-        if str(DataType.Fxp).lower() not in datatype.lower():
-            return False
+    def unpack_data(self, data):
+        data = data & self._data_mask
+        if self._datatype == DataType.Sgl:
+            return ctypes.c_float.from_buffer(ctypes.c_uint(data)).value
+        if self._datatype == DataType.Dbl:
+            return ctypes.c_double.from_buffer(ctypes.c_ulonglong(data)).value
+
+    def pack_data(self, data_to_pack, packed_data):
+        if self._datatype == DataType.Sgl:
+            bits_to_pack = ctypes.c_uint.from_buffer(ctypes.c_float(data_to_pack)).value
+        if self._datatype == DataType.Dbl:
+            bits_to_pack = ctypes.c_ulonglong.from_buffer(ctypes.c_double(data_to_pack)).value
+        return (packed_data << self._size_in_bits) | bits_to_pack
+
+
+class _Bool(object):
+    """ Handles packing and unpacking bools. """
+    def __init__(self, name):
+        self._name = name
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return DataType.Bool
+
+    @property
+    def size_in_bits(self):
+        return 1
+
+    @property
+    def is_c_api_type(self):
         return True
 
+    def unpack_data(self, data):
+        return bool(data & 1)
+
+    def pack_data(self, data_to_pack, packed_data):
+        bit_to_pack = 1 if data_to_pack else 0
+        return (packed_data << 1) | bit_to_pack
+
+
+class ClusterMustContainUniqueNames(RuntimeError):
+    """ For the FPGA Interface Python API, we have chosen to represent clusters
+    as dictionaries.  This has a relatively straight forward conversion, but
+    requires that all members of the cluster have a unique label."""
+    pass
+
+
+class _Cluster(object):
+    """ Handles packing and unpacking clusters. """
+    def __init__(self, name, type_xml):
+        self._name = name
+        self._datatype = DataType.Cluster
+        member_types = type_xml.find("TypeList")
+        self._children = []
+        names = set()
```

Comment:
Ya although it might be overkill since self._children already contains this info and clusters aren't really ever going to be that huge.

I could replace "child_type.name in names" with something like:
    child for child in self._children if child.name == child_type.name

What do you think?